File: js/net/src/compression.ts
/**
 * Per-frame payload compression for moq-lite-05.
 *
 * A publisher marks a {@link Track} with `compress = true` when its frames are
 * worth compressing (e.g. a JSON catalog). The concrete codec is negotiated in
 * SUBSCRIBE_OK, and every frame is compressed independently so the codec never
 * carries state across the lossy, out-of-order group boundary.
 */

// Mirrors the Rust MAX_FRAME_SIZE cap on the receive path: reject anything that
// inflates past this so a malicious peer can't zip-bomb the receiver.
const MAX_FRAME_SIZE = 16 * 1024 * 1024;

/** The codec used to (de)compress frame payloads, negotiated per subscription. */
export const Compression = {
	/** Frames are sent verbatim. */
	None: 0,
	/**
	 * Raw DEFLATE (RFC 1951), no zlib/gzip header. Matches the browser's
	 * "deflate-raw" format and the Rust side's `flate2` raw deflate. QUIC already
	 * guarantees integrity, so the extra checksum bytes of zlib/gzip are wasted.
	 */
	Deflate: 1,
} as const;

export type Compression = (typeof Compression)[keyof typeof Compression];

/** Parse a wire varint code, throwing on an unknown codec. */
export function compressionFromCode(code: number): Compression {
	switch (code) {
		case Compression.None:
			return Compression.None;
		case Compression.Deflate:
			return Compression.Deflate;
		default:
			throw new Error(`unsupported compression codec: ${code}`);
	}
}

// Map a codec to its WHATWG (De)CompressionStream format string.
function format(codec: Compression): CompressionFormat {
	switch (codec) {
		case Compression.Deflate:
			return "deflate-raw";
		default:
			throw new Error(`codec has no stream format: ${codec}`);
	}
}

function concat(chunks: Uint8Array[], total: number): Uint8Array {
	if (chunks.length === 1) return chunks[0];
	const out = new Uint8Array(total);
	let offset = 0;
	for (const chunk of chunks) {
		out.set(chunk, offset);
		offset += chunk.byteLength;
	}
	return out;
}

// Pump `data` through a transform stream and collect the result, capping the
// output at `maxSize` (used on decode so a small compressed blob can't expand
// without bound).
async function pump(transform: CompressionStream | DecompressionStream, data: Uint8Array, maxSize: number) {
	const writer = transform.writable.getWriter();
	// Drive the write concurrently with the read so the transform doesn't stall
	// waiting for its output to be drained.
	const writing = (async () => {
		// Cast per the project convention for passing a Uint8Array to a Web
		// Streams sink typed as BufferSource (see watch/src/audio/mse.ts).
		await writer.write(data as BufferSource);
		await writer.close();
	})();

	const reader = transform.readable.getReader();
	const chunks: Uint8Array[] = [];
	let total = 0;

	try {
		for (;;) {
			const { done, value } = await reader.read();
			if (done) break;
			total += value.byteLength;
			if (total > maxSize) {
				throw new Error(`decompressed frame exceeds ${maxSize} bytes`);
			}
			chunks.push(value);
		}
	} finally {
		// Surface a write-side failure (e.g. malformed deflate input) instead of
		// swallowing it, but don't let an already-thrown read error get masked.
		await writing.catch(() => {});
	}

	return concat(chunks, total);
}

/** Compress a whole frame payload. {@link Compression.None} returns the input unchanged. */
export async function compress(codec: Compression, data: Uint8Array): Promise<Uint8Array> {
	if (codec === Compression.None) return data;
	return pump(new CompressionStream(format(codec)), data, Number.POSITIVE_INFINITY);
}

/**
 * Decompress a whole frame payload, rejecting anything that inflates past
 * `maxSize` (default {@link MAX_FRAME_SIZE}).
 */
export async function decompress(codec: Compression, data: Uint8Array, maxSize = MAX_FRAME_SIZE): Promise<Uint8Array> {
	if (codec === Compression.None) return data;
	return pump(new DecompressionStream(format(codec)), data, maxSize);
}

File: rs/moq-net/src/model/compression.rs
//! Per-frame payload compression.
//!
//! A publisher marks a [`crate::Track`] with `compress = true` when its frames are
//! worth compressing (e.g. a JSON catalog). The wire protocol then negotiates a
//! concrete [`Compression`] codec in SUBSCRIBE_OK, and every frame on that track is
//! compressed independently so the codec doesn't carry state across the lossy,
//! out-of-order group boundary.

use std::io::{Read, Write};

use crate::{Error, MAX_FRAME_SIZE, Result};

/// The codec used to (de)compress frame payloads, negotiated per subscription.
#[derive(Clone, Copy, Debug, Default, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
pub enum Compression {
	/// Frames are sent verbatim.
	#[default]
	None,
	/// Raw DEFLATE (RFC 1951), no zlib/gzip header. QUIC already guarantees
	/// integrity, so the extra checksum bytes of zlib/gzip would be wasted.
	Deflate,
}

impl Compression {
	/// Compress a whole frame payload.
	///
	/// [`Compression::None`] returns the input unchanged. The caller decides
	/// whether the result is actually smaller; this just applies the codec.
	pub fn compress(&self, data: &[u8]) -> Vec<u8> {
		match self {
			Self::None => data.to_vec(),
			Self::Deflate => {
				let mut encoder = flate2::write::DeflateEncoder::new(Vec::new(), flate2::Compression::default());
				// Writing into a Vec is infallible.
				encoder.write_all(data).expect("deflate write to vec");
				encoder.finish().expect("deflate finish to vec")
			}
		}
	}

	/// Decompress a whole frame payload, rejecting anything that inflates past
	/// `MAX_FRAME_SIZE` so a malicious peer can't zip-bomb the receiver.
	pub fn decompress(&self, data: &[u8]) -> Result<Vec<u8>> {
		match self {
			Self::None => Ok(data.to_vec()),
			Self::Deflate => {
				// Read one byte past the limit so we can tell "exactly at the cap"
				// apart from "overflowed".
				let mut decoder = flate2::read::DeflateDecoder::new(data).take(MAX_FRAME_SIZE + 1);
				let mut out = Vec::new();
				decoder.read_to_end(&mut out).map_err(|_| Error::Decompress)?;
				if out.len() as u64 > MAX_FRAME_SIZE {
					return Err(Error::FrameTooLarge);
				}
				Ok(out)
			}
		}
	}

	/// The varint code used on the wire.
	pub fn to_code(self) -> u64 {
		match self {
			Self::None => 0,
			Self::Deflate => 1,
		}
	}

	/// Parse a wire varint code, erroring on unknown codecs.
	pub fn from_code(code: u64) -> Result<Self> {
		match code {
			0 => Ok(Self::None),
			1 => Ok(Self::Deflate),
			_ => Err(Error::Unsupported),
		}
	}
}

#[cfg(test)]
mod test {
	use super::*;

	#[test]
	fn none_roundtrip() {
		let data = b"the quick brown fox";
		let c = Compression::None;
		let packed = c.compress(data);
		assert_eq!(&packed, data);
		assert_eq!(c.decompress(&packed).unwrap(), data);
	}

	#[test]
	fn deflate_roundtrip() {
		// Highly compressible input so we can assert the codec actually shrinks it.
		let data = vec![b'a'; 4096];
		let c = Compression::Deflate;
		let packed = c.compress(&data);
		assert!(packed.len() < data.len(), "deflate should shrink repetitive data");
		assert_eq!(c.decompress(&packed).unwrap(), data);
	}

	#[test]
	fn deflate_empty() {
		let c = Compression::Deflate;
		let packed = c.compress(&[]);
		assert_eq!(c.decompress(&packed).unwrap(), Vec::<u8>::new());
	}

	#[test]
	fn decompress_rejects_garbage() {
		let c = Compression::Deflate;
		assert!(matches!(c.decompress(b"not a deflate stream"), Err(Error::Decompress)));
	}

	#[test]
	fn code_roundtrip() {
		for c in [Compression::None, Compression::Deflate] {
			assert_eq!(Compression::from_code(c.to_code()).unwrap(), c);
		}
		assert!(Compression::from_code(99).is_err());
	}
}
